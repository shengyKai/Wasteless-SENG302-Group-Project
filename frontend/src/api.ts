/*
 * Created on Wed Feb 10 2021
 *
 * The Unlicense
 * This is free and unencumbered software released into the public domain.
 *
 * Anyone is free to copy, modify, publish, use, compile, sell, or distribute
 * this software, either in source code form or as a compiled binary, for any
 * purpose, commercial or non-commercial, and by any means.
 *
 * In jurisdictions that recognize copyright laws, the author or authors of this
 * software dedicate any and all copyright interest in the software to the public
 * domain. We make this dedication for the benefit of the public at large and to
 * the detriment of our heirs and successors. We intend this dedication to be an
 * overt act of relinquishment in perpetuity of all present and future rights to
 * this software under copyright law.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 * For more information, please refer to <https://unlicense.org>
 */

/**
 * Declare all available services here
 */
import axios from 'axios';
const SERVER_URL = process.env.VUE_APP_SERVER_ADD;

const instance = axios.create({
  baseURL: SERVER_URL,
  timeout: 2000,
  withCredentials: true,
});

type MaybeError<T> = T | string;

export type User = {
  id: number,
  firstName: string,
  lastName: string,
  middleName?: string,
  nickname?: string,
  bio?: string,
  email: string,
  dateOfBirth: string,
  phoneNumber?: string,
  homeAddress: Location,
  created?: string,
  role?: "user" | "globalApplicationAdmin" | "defaultGlobalApplicationAdmin",
  businessesAdministered?: Business[],
};

export type Location = {
  streetNumber?: string,
  streetName?: string,
  city?: string,
  region?: string,
  country: string,
  postcode?: string,
};

export type CreateUser = {
  firstName: string,
  lastName: string,
  middleName?: string,
  nickname?: string,
  bio?: string,
  email: string,
  dateOfBirth: string,
  phoneNumber?: string,
  homeAddress: Location,
  password: string,
};

export type BusinessType = 'Accommodation and Food Services' | 'Retail Trade' | 'Charitable organisation' | 'Non-profit organisation';

export const BUSINESS_TYPES: BusinessType[] = ['Accommodation and Food Services', 'Retail Trade', 'Charitable organisation', 'Non-profit organisation'];

export type Business = {
  id: number,
  administrators?: User[],
  name: string,
  description?: string,
  address: string,
  businessType: BusinessType,
  created?: string,
};

export type CreateBusiness = {
  name: string,
  description?: string,
  address: string,
  businessType: BusinessType,
};

export type Image = {
  id: number,
  filename: string,
  thumbnailFilename: string,
};

export type Product = {
  id: string,
  name: string,
  description?: string,
  manufacturer?: string,
  recommendedRetailPrice?: number,
  created?: string,
  images: Image[],
};

export type CreateProduct = Omit<Product, 'created' | 'images'>;

function isLocation(obj: any): obj is Location {
  if (obj === null || typeof obj !== 'object') return false;
  if (obj.streetNumber !== undefined && typeof obj.streetNumber !== 'string') return false;
  if (obj.streetName !== undefined && typeof obj.streetName !== 'string') return false;
  if (obj.city !== undefined && typeof obj.city !== 'string') return false;
  if (obj.region !== undefined && typeof obj.region !== 'string') return false;
  if (typeof obj.country !== 'string') return false;
  if (obj.postcode !== undefined && typeof obj.postcode !== 'string') return false;

  return true;
}

function isUser(obj: any): obj is User {
  if (obj === null || typeof obj !== 'object') return false;
  if (typeof obj.id !== 'number') return false;
  if (typeof obj.firstName !== 'string') return false;
  if (typeof obj.lastName !== 'string') return false;
  if (obj.middleName !== undefined && typeof obj.middleName !== 'string') return false;
  if (obj.nickname !== undefined && typeof obj.nickname !== 'string') return false;
  if (obj.bio !== undefined && typeof obj.bio !== 'string') return false;
  if (typeof obj.email !== 'string') return false;
  if (typeof obj.dateOfBirth !== 'string') return false;
  if (obj.phoneNumber !== undefined && typeof obj.phoneNumber !== 'string') return false;
  if (!isLocation(obj.homeAddress)) return false;
  if (obj.created !== undefined && typeof obj.created !== 'string') return false;
  if (obj.role !== undefined && !['user', 'globalApplicationAdmin', 'defaultGlobalApplicationAdmin'].includes(obj.role))
    return false;
  if (obj.businessesAdministered !== undefined && !isBusinessArray(obj.businessesAdministered)) return false;

  return true;
}

function isBusiness(obj: any): obj is Business {
  if (obj === null || typeof obj !== 'object') return false;
  if (typeof obj.id !== 'number') return false;
  if (obj.administrators !== undefined && !isUserArray(obj.administrators)) return false;
  if (typeof obj.name !== 'string') return false;
  if (obj.description !== undefined && typeof obj.description !== 'string') return false;
  if (typeof obj.address !== 'object') return false;
  if (!BUSINESS_TYPES.includes(obj.businessType)) return false;
  if (obj.created !== undefined && typeof obj.created !== 'string') return false;

  return true;
}

function isNumberArray(obj: any): obj is number[] {
  if (!Array.isArray(obj)) return false;
  for (let elem of obj) {
    if (typeof elem !== 'number') return false;
  }
  return true;
}

function isUserArray(obj: any): obj is User[] {
  if (!Array.isArray(obj)) return false;
  for (let elem of obj) {
    if (!isUser(elem)) return false;
  }
  return true;
}

function isBusinessArray(obj: any): obj is Business[] {
  if (!Array.isArray(obj)) return false;
  for (let elem of obj) {
    if (!isBusiness(elem)) return false;
  }
  return true;
}

type OrderBy = 'userId' | 'relevance' | 'firstName' | 'middleName' | 'lastName' | 'nickname' | 'email' | 'address';

/**
 * Sends a search query to the backend.
 *
 * @param query Query string to search for
 * @param pageIndex Index of page to start the results from (1 = first page)
 * @param resultsPerPage Number of results to return per page
 * @param orderBy Specifies the method used to sort the results
 * @param reverse Specifies whether to reverse the search results (default order is descending for relevance and ascending for all other orders)
 * @returns List of user infos for the current page or an error message
 */
export async function search(query: string, pageIndex: number, resultsPerPage: number, orderBy: OrderBy, reverse: boolean): Promise<MaybeError<User[]>> {
  let response;
  try {
    response = await instance.get('/users/search', {
      params: {
        searchQuery: query,
        page: pageIndex,
        resultsPerPage,
        orderBy,
        reverse: reverse.toString(),
      }
    });
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    return `Request failed: ${status}`;
  }

  if (!isUserArray(response.data)) {
    return 'Response is not user array';
  }

  return response.data;
}

/**
 * Sends a query for the number of search results for a given query to the backend.
 *
 * @param query Query string to search for
 * @returns Number of search results or an error message
 */
export async function getSearchCount(query: string): Promise<MaybeError<number>> {
  let response;
  try {
    response = await instance.get('/users/search/count', {
      params: {
        searchQuery: query,
      }
    });
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    return `Request failed: ${status}`;
  }

  if (typeof response.data?.count !== 'number') {
    return 'Response is not number';
  }

  return response.data.count;
}

/**
 * Queries the backend for a specific user by their id.
 *
 * @param id User id, if ommitted then fetch the logged in user's info
 * @returns User info for the given id or an error message
 */
export async function getUser(id: number): Promise<MaybeError<User>> {
  let response;
  try {
    response = await instance.get('/users/' + id);
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    return `Request failed: ${status}`;
  }

  if (!isUser(response.data)) {
    return 'Response is not user';
  }

  return response.data;
}

/**
 * Logs in to the given user account by setting the authentication cookie.
 *
 * @param email User email
 * @param password User password
 * @returns The now logged in user ID if operation is successful, otherwise a string error.
 */
export async function login(email: string, password: string): Promise<MaybeError<number>> {
  let response;
  try {
    response = await instance.post('/login', {
      email: email,
      password: password,
    });
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    if (status === 400) return 'Invalid credentials';
    return `Request failed: ${status}`;
  }
  let id = response.data.userId;
  if (typeof id !== 'number') return 'Invalid response';

  return id;
}

/**
 * Creates a user with the given properties.
 * Note: This doesn't automatically log the user in.
 *
 * @param user Initial user properties
 * @returns undefined if operation is successful, otherwise a string error.
 */
export async function createUser(user: CreateUser): Promise<MaybeError<undefined>> {
  try {
    await instance.post('/users', user);
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    if (status === 400) return 'Invalid create user request';
    if (status === 409) return 'Email in use';
    return 'Request failed: ' + status;
  }

  return undefined;
}

/**
 * Makes the given user a GAA, if the current user is a DGAA.
 *
 * @param userId User to make GAA
 * @returns undefined if operation is successful, otherwise a string error.
 */
export async function makeAdmin(userId: number): Promise<MaybeError<undefined>> {
  try {
    await instance.post(`/users/${userId}/makeAdmin`);
  } catch (error) {
    let status: number | undefined = error.response?.status;

    if (status === undefined) return 'Failed to reach backend';
    if (status === 401) return 'Missing/Invalid access token';
    if (status === 403) return 'Operation not permitted';
    if (status === 406) return 'User does not exist';

    return 'Request failed: ' + status;
  }
  return undefined;
}

/**
 * Revokes the given user's admin rights
 *
 * @param userId User to revoke permissions
 * @returns undefined if operation is successful, otherwise a string error.
 */
export async function revokeAdmin(userId: number): Promise<MaybeError<undefined>> {
  try {
    await instance.post(`/users/${userId}/revokeAdmin`);
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === 401) return 'Missing/Invalid access token';
    if (status === 403) return 'Operation not permitted';
    if (status === 406) return 'User does not exist';

    return 'Request failed: ' + status;
  }
  return undefined;
}

/**
 * Creates a business
 *
 * @param business The properties to create the business with
 * @returns undefined if operation is successful, otherwise a string error.
 */
export async function createBusiness(business: CreateBusiness): Promise<MaybeError<undefined>> {
  try {
    await instance.post('/businesses', business);
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 401) return 'Missing/Invalid access token';

    return 'Request failed: ' + status;
  }

  return undefined;
}

/**
 * Add a product to a businesses catalogue.
 *
 * @param businessId The business to add the product to
 * @param product The properties to create a product with
 * @return undefined if operation is successful, otherwise a string error
 */
export async function createProduct(businessId: number, product: CreateProduct): Promise<MaybeError<undefined>> {
  try {
    await  instance.post(`/businesses/${businessId}/products`, product);
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 401) return 'Missing/Invalid access token';
    if (status === 403) return 'Operation not permitted';

    if (status === 400) {
      // TODO Not sure exactly how the backend is going to communicate with us that the product code
      // is unavailable.

      // eslint-disable-next-line no-constant-condition
      if (false) {
        return 'Product code unavailable';
      }

      return 'Invalid parameters';
    }

    return 'Request failed: ' + status;
  }
  return undefined;
}

/**
 * Fetches a business with the given id.
 *
 * @param businessId Business id to fetch
 * @returns The requested business if operation is successful, otherwise a string error.
 */
export async function getBusiness(businessId: number): Promise<MaybeError<Business>> {
  let response;
  try {
    response = await instance.get(`/businesses/${businessId}`);
  } catch (error) {
    let status: number | undefined = error.response?.status;
    if (status === undefined) return 'Failed to reach backend';
    if (status === 401) return 'Missing/Invalid access token';
    if (status === 406) return 'Business not found';

    return 'Request failed: ' + status;
  }

  if (!isBusiness(response.data)) {
    return 'Invalid response type';
  }

  return response.data;
}