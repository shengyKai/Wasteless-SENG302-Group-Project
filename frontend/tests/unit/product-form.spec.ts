import Vue from 'vue';
import Vuetify from 'vuetify';
import {createLocalVue, mount, Wrapper} from '@vue/test-utils';

import ProductForm from '@/components/BusinessProfile/ProductForm.vue';
import {castMock, flushQueue, findButtonWithText} from './utils';
import {createProduct as createProduct1, modifyProduct as modifyProduct1, Product} from "@/api/internal-product";
import { currencyFromCountry } from '@/api/currency';

jest.mock('@/api/internal', () => ({
  createProduct: jest.fn(),
  modifyProduct: jest.fn(),
  getBusiness: jest.fn(() => {
    return {
      address: {
        country: 'New Zealand',
      }
    };
  })
}));

jest.mock('@/api/currency', () => ({
  currencyFromCountry: jest.fn(() => {
    return {
      code: 'Currency code',
      name: 'Currency name',
      symbol: 'Currency symbol',
      errorMessage: "Some error message"
    };
  })
}));


const createProduct = castMock(createProduct1);
const modifyProduct = castMock(modifyProduct1);

Vue.use(Vuetify);

// Characters that are in the set of letters, numbers, spaces and punctuation.
const validCharacters = [" ", ":", ",", "7", "é", "树", "A"];
// Characters that are not a letter, number, space or punctuation.
const invalidCharacters = ["\uD83D\uDE02", "♔"];
// Characters that are whitespace not including the space character.
const whitespaceCharacters = ["\n", "\t"];

const localVue = createLocalVue();

describe('ProductForm.vue - Create', () => {
  // Container for the wrapper around ProductForm
  let appWrapper: Wrapper<any>;

  // Container for the ProductForm under test
  let wrapper: Wrapper<any>;

  /**
   * Sets up the test ProductForm instance
   *
   * Because the element we're testing has a v-dialog we need to take some extra sets to make it
   * work.
   */
  beforeEach(() => {
    const vuetify = new Vuetify();

    // Creating wrapper around ProductForm with data-app to appease vuetify
    const App = localVue.component('App', {
      components: { ProductForm },
      template: '<div data-app><ProductForm :previousProduct="previousProduct" :businessId="90"/></div>',
      data() {
        return {
          previousProduct: undefined,
        };
      }
    });

    // Put the ProductForm component inside a div in the global document,
    // this seems to make vuetify work correctly, but necessitates calling appWrapper.destroy
    const elem = document.createElement('div');
    document.body.appendChild(elem);

    appWrapper = mount(App, {
      localVue,
      vuetify,
      attachTo: elem,
    });

    wrapper = appWrapper.getComponent(ProductForm);
  });

  /**
   * Executes after every test case.
   *
   * This function makes sure that the ProductForm component is removed from the global document
   */
  afterEach(() => {
    appWrapper.destroy();
  });

  /**
   * Adds all the fields that are required for the create business form to be valid
   *
   * These are:
   * - Product name
   * - Product shortcode
   */
  async function populateRequiredFields() {
    await wrapper.setData({
      product: 'Product Name',
      productCode: 'ABC-XYZ-012-789',
    });
  }

  /**
   * Populates all fields of the ProductForm form
   *
   * Which include the product's:
   *  - Name
   *  - Shortcode
   *  - Description
   *  - Manufacturer
   *  - Recommended retail price
   */
  async function populateAllFields() {
    await populateRequiredFields();
    await wrapper.setData({
      description: 'Product Description',
      manufacturer: 'Product Manufacturer',
      recommendedRetailPrice: '100.00',
    });
  }

  /**
   * Finds the close button in the ProductForm form
   *
   * @returns A Wrapper around the close button
   */
  const findCloseButton = () => findButtonWithText(wrapper, 'Close');

  /**
   * Finds the create button in the ProductForm form
   *
   * @returns A Wrapper around the create button
   */
  const findCreateButton = () => findButtonWithText(wrapper, 'Create');

  it('Valid if all required fields are provided', async () => {
    await populateRequiredFields();
    await Vue.nextTick();
    expect(wrapper.vm.valid).toBeTruthy();
  });

  it('Valid if all fields are provided', async () => {
    await populateAllFields();
    await Vue.nextTick();
    expect(wrapper.vm.valid).toBeTruthy();
  });

  it.each(validCharacters)('Valid if product has name "%s"', async (name) => {
    await populateAllFields();
    await wrapper.setData({
      product: name,
    });
    await Vue.nextTick();
    expect(wrapper.vm.valid).toBeTruthy();
  });

  it.each(invalidCharacters.concat(whitespaceCharacters))(`Invalid if product has name "%s"`, async (name) => {
    await populateAllFields();
    await wrapper.setData({
      product: name,
    });
    await Vue.nextTick();
    expect(wrapper.vm.valid).toBeFalsy();
  });

  it('Invalid if no product name', async () => {
    await populateRequiredFields();
    await wrapper.setData({
      product: '',
    });
    await Vue.nextTick();
    expect(wrapper.vm.valid).toBeFalsy();
  });

  it('Invalid if product name is too long', async () => {
    await populateRequiredFields();
    await wrapper.setData({
      product: 'a'.repeat(101),
    });
    await Vue.nextTick();
    expect(wrapper.vm.valid).toBeFalsy();
  });

  it.each(validCharacters.concat(whitespaceCharacters))('Valid if product has description "%s"', async (description) => {
    await populateAllFields();
    await wrapper.setData({
      description,
    });
    await Vue.nextTick();
    expect(wrapper.vm.valid).toBeTruthy();
  });

  it.each(invalidCharacters)('Invalid if product has description "%s"', async (description) => {
    await populateAllFields();
    await wrapper.setData({
      description,
    });
    await Vue.nextTick();
    expect(wrapper.vm.valid).toBeFalsy();
  });

  it('Invalid if description is too long', async () => {
    await populateRequiredFields();
    await wrapper.setData({
      description: 'a'.repeat(201),
    });
    await Vue.nextTick();
    expect(wrapper.vm.valid).toBeFalsy();
  });

  it.each(validCharacters)('Valid if product has manufacturer "%s"', async (manufacturer) => {
    await populateAllFields();
    await wrapper.setData({
      manufacturer,
    });
    await Vue.nextTick();
    expect(wrapper.vm.valid).toBeTruthy();
  });

  it.each(invalidCharacters.concat(whitespaceCharacters))('Invalid if product has manufacturer "%s"', async (manufacturer) => {
    await populateAllFields();
    await wrapper.setData({
      manufacturer,
    });
    await Vue.nextTick();
    expect(wrapper.vm.valid).toBeFalsy();
  });

  it('Invalid if manufacturer is too long', async () => {
    await populateRequiredFields();
    await wrapper.setData({
      manufacturer: 'a'.repeat(101),
    });
    await Vue.nextTick();
    expect(wrapper.vm.valid).toBeFalsy();
  });

  it('Invalid if product price is not a number', async () => {
    await populateRequiredFields();
    await wrapper.setData({
      recommendedRetailPrice: 'not a number',
    });
    await Vue.nextTick();
    expect(wrapper.vm.valid).toBeFalsy();
  });

  it('Invalid if product price is negative', async () => {
    await populateRequiredFields();
    await wrapper.setData({
      recommendedRetailPrice: '-10',
    });
    await Vue.nextTick();
    expect(wrapper.vm.valid).toBeFalsy();
  });

  it('Invalid if product price is too high', async () => {
    await populateRequiredFields();
    await wrapper.setData({
      recommendedRetailPrice: '100001',
    });
    await Vue.nextTick();
    expect(wrapper.vm.valid).toBeFalsy();
  });

  it('Invalid if product price has too many digits after the decimal', async () => {
    await populateRequiredFields();
    await wrapper.setData({
      recommendedRetailPrice: '3.141',
    });
    await Vue.nextTick();
    expect(wrapper.vm.valid).toBeFalsy();
  });

  it('Invalid if product price has a single decimal digit', async () => {
    await populateRequiredFields();
    await wrapper.setData({
      recommendedRetailPrice: '3.1',
    });
    await Vue.nextTick();
    expect(wrapper.vm.valid).toBeFalsy();
  });

  it('Invalid if no product code is provided', async () => {
    await populateRequiredFields();
    await wrapper.setData({
      productCode: '',
    });
    await Vue.nextTick();
    expect(wrapper.vm.valid).toBeFalsy();
  });

  it('Invalid if product code is too long', async () => {
    await populateRequiredFields();
    await wrapper.setData({
      productCode: 'a'.repeat(16),
    });
    await Vue.nextTick();
    expect(wrapper.vm.valid).toBeFalsy();
  });

  it('Invalid if product code uses lowercase letters', async () => {
    await populateRequiredFields();
    await wrapper.setData({
      productCode: 'aaaaaaaaa',
    });
    await Vue.nextTick();
    expect(wrapper.vm.valid).toBeFalsy();
  });

  it('Invalid if product code has spaces', async () => {
    await populateRequiredFields();
    await wrapper.setData({
      productCode: '          ',
    });
    await Vue.nextTick();
    expect(wrapper.vm.valid).toBeFalsy();
  });

  it('When the close button is pressed then the "closeDialog" event should be emitted', async () => {
    await findCloseButton().trigger('click'); // Click close button
    expect(wrapper.emitted().closeDialog).toBeTruthy();
  });

  it('When the create button is pressed then an api call should be made and is successful', async () => {
    await populateAllFields();
    createProduct.mockResolvedValue(undefined); // Ensure that the operation is successful
    await Vue.nextTick();
    await findCreateButton().trigger('click'); // Click create button
    await Vue.nextTick();
    expect(createProduct).toBeCalledWith(90, {
      id: 'ABC-XYZ-012-789',
      name: 'Product Name',
      description: 'Product Description',
      manufacturer: 'Product Manufacturer',
      recommendedRetailPrice: 100,
    });
    expect(wrapper.emitted().closeDialog).toBeTruthy(); // The dialog should close
  });

  it('When the create button is pressed and the api returns an error then the error should be shown', async () => {
    await populateAllFields();
    createProduct.mockResolvedValue('test_error_message'); // Ensure that the operation fails
    await Vue.nextTick();
    await findCreateButton().trigger('click'); // Click create button
    await flushQueue();
    // The appWrapper is tested for the text, because the dialog content is not in the dialog
    // element.
    expect(appWrapper.text()).toContain('test_error_message');
    expect(wrapper.emitted().closeDialog).toBeFalsy(); // The dialog should stay open
  });

  it('When the create button is pressed and the api says that the product code is unavailable then the form should become invalid', async () => {
    await populateAllFields();
    createProduct.mockResolvedValue('Product code unavailable'); // Ensure that the operation fails
    await Vue.nextTick();
    await findCreateButton().trigger('click'); // Click create button
    await flushQueue();
    expect(wrapper.vm.valid).toBeFalsy();
    expect(wrapper.emitted().closeDialog).toBeFalsy(); // The dialog should stay open
  });

  it('RRP field contains currency code and symbol recieved from API', () => {
    const fields = wrapper.findAllComponents({ name: 'v-text-field' });
    const rrpFields = fields.filter(field => field.text().includes('Recommended Retail Price'));
    expect(rrpFields.length).toBe(1);
    const rrpField = rrpFields.at(0);
    expect(rrpField.text()).toContain('Currency symbol');
    expect(rrpField.text()).toContain('Currency code');
  });

  it('If the error message is not undefined, there should be an hint message that appears in the dialog box', async () => {
    expect(wrapper.vm.currency.errorMessage).toBe("Some error message");
  });
});


describe('ProductForm.vue - Modify', () => {
  // Container for the wrapper around ProductForm
  let appWrapper: Wrapper<any>;

  // Container for the ProductForm under test
  let wrapper: Wrapper<any>;

  const previousProduct: Product = {
    id: 'TEST-ID',
    name: 'Test product name',
    description: 'Test product description',
    manufacturer: 'Test product manufacturer',
    recommendedRetailPrice: 199,
    images: [],
  };

  /**
   * Sets up the test ProductForm instance
   *
   * Because the element we're testing has a v-dialog we need to take some extra sets to make it
   * work.
   */
  beforeEach(() => {
    const vuetify = new Vuetify();

    // Creating wrapper around ProductForm with data-app to appease vuetify
    const App = localVue.component('App', {
      components: { ProductForm },
      template: '<div data-app><ProductForm :previousProduct="previousProduct" :businessId="90"/></div>',
      data() {
        return {
          previousProduct,
        };
      }
    });

    // Put the ProductForm component inside a div in the global document,
    // this seems to make vuetify work correctly, but necessitates calling appWrapper.destroy
    const elem = document.createElement('div');
    document.body.appendChild(elem);

    appWrapper = mount(App, {
      localVue,
      vuetify,
      attachTo: elem,
    });

    wrapper = appWrapper.getComponent(ProductForm);
  });

  /**
   * Executes after every test case.
   *
   * This function makes sure that the ProductForm component is removed from the global document
   */
  afterEach(() => {
    appWrapper.destroy();
  });

  /**
   * Finds the save button in the ProductForm form
   *
   * @returns A Wrapper around the save button
   */
  const findSaveButton = () => findButtonWithText(wrapper, 'Save');

  it('Form has the correct title', () => {
    expect(appWrapper.text()).toContain(`Update ${previousProduct.id}`);
  });

  it('Product fields are populated', () => {
    expect(wrapper.vm.productCode).toBe(previousProduct.id);
    expect(wrapper.vm.product).toBe(previousProduct.name);
    expect(wrapper.vm.description).toBe(previousProduct.description);
    expect(wrapper.vm.manufacturer).toBe(previousProduct.manufacturer);
  });

  it('When the save button is pressed then an api call should be made and is successful', async () => {
    modifyProduct.mockResolvedValue(undefined); // Ensure that the operation is successful
    await wrapper.setData({
      productCode: 'FOO-BAR',
    });
    await Vue.nextTick();

    await findSaveButton().trigger('click'); // Press save

    await flushQueue();

    expect(modifyProduct).toBeCalledWith(90, previousProduct.id, {
      id: 'FOO-BAR',
      name: previousProduct.name,
      description: previousProduct.description,
      manufacturer: previousProduct.manufacturer,
      recommendedRetailPrice: previousProduct.recommendedRetailPrice,
    });
    expect(wrapper.emitted().closeDialog).toBeTruthy(); // The dialog should close
  });

  it('When the save button is pressed and the api returns an error then the error should be shown', async () => {
    modifyProduct.mockResolvedValue('test_error_message'); // Ensure that the operation fails
    await findSaveButton().trigger('click'); // Click save button
    await flushQueue();
    // The appWrapper is tested for the text, because the dialog content is not in the dialog
    // element.
    expect(appWrapper.text()).toContain('test_error_message');
    expect(wrapper.emitted().closeDialog).toBeFalsy(); // The dialog should stay open
  });

  it('When the save button is pressed and the api says that the product code is unavailable then the form should become invalid', async () => {
    modifyProduct.mockResolvedValue('Product code unavailable'); // Ensure that the operation fails
    await findSaveButton().trigger('click'); // Click save button
    await flushQueue();
    expect(wrapper.vm.valid).toBeFalsy();
    expect(wrapper.emitted().closeDialog).toBeFalsy(); // The dialog should stay open
  });
});

