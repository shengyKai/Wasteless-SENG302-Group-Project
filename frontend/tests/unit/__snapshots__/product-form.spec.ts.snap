// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductForm.vue - Modify Matches snapshot 1`] = `
<div data-app="">
  <div class="v-overlay theme--dark" style="z-index: 5;">
    <div class="v-overlay__scrim" style="opacity: 0; background-color: rgb(33, 33, 33); border-color: #212121;"></div>
  </div>
  <div class="row justify-center">
    <div role="dialog" class="v-dialog__container"></div>
  </div>
  <div role="document" tabindex="0" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div class="v-dialog v-dialog--active v-dialog--persistent" style="transform-origin: center center; max-width: 600px;">
        <form novalidate="novalidate" class="v-form">
          <div class="v-card v-sheet theme--light">
            <div class="v-card__title"><span class="headline create-product">
              Update TEST-ID
            </span></div>
            <div class="v-card__text">
              <div class="container">
                <div class="row">
                  <div class="col">
                    <div class="v-input required v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
                      <div class="v-input__control">
                        <div class="v-input__slot">
                          <fieldset aria-hidden="true">
                            <legend style="width: 0px;"><span>​</span></legend>
                          </fieldset>
                          <div class="v-text-field__slot"><label for="input-1467" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Short-hand Product Code</label><input id="input-1467" type="text"></div>
                        </div>
                        <div class="v-text-field__details">
                          <div class="v-messages theme--light">
                            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div class="col">
                    <div class="v-input required v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
                      <div class="v-input__control">
                        <div class="v-input__slot">
                          <fieldset aria-hidden="true">
                            <legend style="width: 0px;"><span>​</span></legend>
                          </fieldset>
                          <div class="v-text-field__slot"><label for="input-1470" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Name of product</label><input id="input-1470" type="text"></div>
                        </div>
                        <div class="v-text-field__details">
                          <div class="v-messages theme--light">
                            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div class="col col-12">
                    <div class="v-input v-textarea v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
                      <div class="v-input__control">
                        <div class="v-input__slot">
                          <fieldset aria-hidden="true">
                            <legend style="width: 0px;"><span>​</span></legend>
                          </fieldset>
                          <div class="v-text-field__slot"><label for="input-1473" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Description</label><textarea id="input-1473" rows="3"></textarea></div>
                        </div>
                        <div class="v-text-field__details">
                          <div class="v-messages theme--light">
                            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div class="col col-6">
                    <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
                      <div class="v-input__control">
                        <div class="v-input__slot">
                          <fieldset aria-hidden="true">
                            <legend style="width: 0px;"><span>​</span></legend>
                          </fieldset>
                          <div class="v-text-field__slot"><label for="input-1476" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Manufacturer</label><input id="input-1476" type="text"></div>
                        </div>
                        <div class="v-text-field__details">
                          <div class="v-messages theme--light">
                            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div class="col col-6">
                    <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--prefix v-text-field--enclosed v-text-field--outlined">
                      <div class="v-input__control">
                        <div class="v-input__slot">
                          <fieldset aria-hidden="true">
                            <legend style="width: 0px;"><span>​</span></legend>
                          </fieldset>
                          <div class="v-text-field__slot"><label for="input-1479" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Recommended Retail Price</label>
                            <div class="v-text-field__prefix">Currency symbol</div><input id="input-1479" type="text">
                            <div class="v-text-field__suffix">Currency code</div>
                          </div>
                        </div>
                        <div class="v-text-field__details">
                          <div class="v-messages theme--light">
                            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div class="v-card__actions">
              <div class="spacer"></div>
              <!---->
              <div class="spacer"></div> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content">
            Close
          </span></button> <button type="submit" class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default primary"><span class="v-btn__content">
              Save
            </span></button>
            </div>
          </div>
        </form>
      </div>
    </transition-stub>
  </div>
</div>
`;
